File: leetcode/python/54_spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix: list[list[int]]) -> list[int]:
        res: list[int] = []
        xl, xr, yl, yr = 0, len(matrix[0]) - 1, 0, len(matrix) - 1
        while xl <= xr and yl <= yr:
            res.extend(matrix[yl][xl:xr + 1])
            yl += 1

            if yl > yr:
                break
            res.extend([matrix[i][xr] for i in range(yl, yr + 1)])
            xr -= 1

            if xl > xr:
                break
            res.extend(matrix[yr][xl:xr+1][::-1])
            yr -= 1

            if yl > yr:
                break
            res.extend([matrix[i][xl] for i in range(yr, yl-1, -1)])
            xl += 1
        return res

s = Solution()
print(s.spiralOrder(matrix = [[1,2,3],[4,5,6],[7,8,9]]))
print(s.spiralOrder(matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]))

File: leetcode/rust/src/54_spiral_matrix.rs
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let mut m = matrix.len();
        let mut n = matrix[0].len();
        let mut res = vec![0; m * n];
        let mut i = 0;
        let mut j = 0;
        let mut k = 0;
        let mut l = 0;
        let mut idx = 0;
        while i < m && j < n {
            for k in j..n {
                res[idx] = matrix[i][k];
                idx += 1;
            }
            i += 1;
            for k in i..m {
                res[idx] = matrix[k][n - 1];
                idx += 1;
            }
            n -= 1;
            if i < m {
                for k in (j..n).rev() {
                    res[idx] = matrix[m - 1][k];
                    idx += 1;
                }
                m -= 1;
            }
            if j < n {
                for k in (i..m).rev() {
                    res[idx] = matrix[k][j];
                    idx += 1;
                }
                j += 1;
            }
        }
        res
    }
}